### PR TITLE
boards: arm: Add support for SiLabs EFM32 Giant Gecko GG11 STK

### PR DESCRIPTION
**Pull request by Oanerer**
This PR adds initial support for the Silicon Labs EFM32 Giant Gecko GG11 Starter Kit (EFM32GG-STK3701A).

Peripherals supported for this board are currently NVIC, SysTick, GPIO, all UARTs/USARTs/LEUARTs, I2C, Flash, Counter and Ethernet. ADC and Watchdog support will be added as soon as their respective PRs are merged. 

The following samples were tested on this board:
* hello_world
* nvs
* counter
* big_http_download
... and a variety of other self written applications.

Note: this PR depends on a PR [#2](https://github.com/zephyrproject-rtos/hal_silabs/pull/2) for hal_silabs. A west.yml entry will be added to the PR when it has been merged.

### CODE REVIEW

File: boards/arm/efm32gg_stk3701a/doc/index.rst
Comment (dbkinder) [inline boards/arm/efm32gg_stk3701a/doc/index.rst]: Interesting that you picked an example "PE2" that's not an entry in the table.

File: boards/arm/efm32gg_stk3701a/doc/index.rst
Comment (dbkinder) [inline boards/arm/efm32gg_stk3701a/doc/index.rst]: Do we need to know which UART is assigned to the console?

File: boards/arm/efm32gg_stk3701a/doc/index.rst
Comment (dbkinder) [inline boards/arm/efm32gg_stk3701a/doc/index.rst]: Do you need to clarify which pins on the header to connect to?

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/efm32gg_stk3701a/doc/index.rst
Comment (Oanerer) [reply boards/arm/efm32gg_stk3701a/doc/index.rst]: True, I'm afraid this is a case of 'copy/paste, no time to waste'. Modified to PE1, which *is* actually used on the board.

File: boards/arm/efm32gg_stk3701a/doc/index.rst
Comment (Oanerer) [reply boards/arm/efm32gg_stk3701a/doc/index.rst]: Added (USART4 in this case)

File: boards/arm/efm32gg_stk3701a/doc/index.rst
Comment (Oanerer) [reply boards/arm/efm32gg_stk3701a/doc/index.rst]: This is definately a case of 'copy/paste'. The console is assigned to USART4, which is exposed through the USB connector along with the mass storage device. So corrected that and removed any mention of the USB-to-UART converter.

**Review by Oanerer — COMMENTED** [review]
David, thanks for your time reviewing. If you feel anything is lacking in general with regards to the documentation please let me know!

**Review by dbkinder — APPROVED** [review]
docs LGTM, thanks.

File: drivers/ethernet/eth_gecko.c
Comment (jukkar) [inline drivers/ethernet/eth_gecko.c]: This is old way of setting up logging. Instead of this, you could just write
```
#include <logging/log.h>
LOG_MODULE_REGISTER(eth_gecko, CONFIG_ETHERNET_LOG_LEVEL);
```

File: drivers/ethernet/eth_gecko.c
Comment (jukkar) [inline drivers/ethernet/eth_gecko.c]: Indentation wrong here

File: drivers/ethernet/eth_gecko.c
Comment (jukkar) [inline drivers/ethernet/eth_gecko.c]: You mentioned in the beginning of the file that statistics are not supported. To me it looks quite trivial task to support that as you only need the interface pointer for it.

File: drivers/ethernet/phy_gecko.c
Comment (jukkar) [inline drivers/ethernet/phy_gecko.c]: This could also shortened to
```
#include <logging/log.h>
LOG_MODULE_REGISTER(eth_gecko_phy, CONFIG_ETHERNET_LOG_LEVEL);
```
or you could just use the eth_gecko module too, in which case this could be written as
```
#include <logging/log.h>
LOG_MODULE_DECLARE(eth_gecko, CONFIG_ETHERNET_LOG_LEVEL);
```

File: drivers/ethernet/phy_gecko.c
Comment (jukkar) [inline drivers/ethernet/phy_gecko.c]: The prints seem to be wrong order, speed should be first followed by duplex setting.

**Review by jukkar — CHANGES_REQUESTED** [review]
Minor nit picks for the ethernet driver

File: drivers/ethernet/eth_gecko.c
Comment (Oanerer) [reply drivers/ethernet/eth_gecko.c]: Changed

File: drivers/ethernet/eth_gecko.c
Comment (Oanerer) [reply drivers/ethernet/eth_gecko.c]: Corrected

File: drivers/ethernet/phy_gecko.c
Comment (Oanerer) [reply drivers/ethernet/phy_gecko.c]: Changed to the first suggestion

File: drivers/ethernet/phy_gecko.c
Comment (Oanerer) [reply drivers/ethernet/phy_gecko.c]: Oops, corrected

File: drivers/ethernet/eth_gecko.c
Comment (Oanerer) [reply drivers/ethernet/eth_gecko.c]: I actually hadn't taken a good look yet at what supporting statistics would require. I uncommented line 232 in the section above, is that enough for this driver or are there other things I should support? If so, could you point me to a driver that I can use as a starting point?

**Review by Oanerer — COMMENTED** [review]
Thanks Jukka for your time reviewing, much appreciated!

File: drivers/ethernet/eth_gecko.c
Comment (jukkar) [reply drivers/ethernet/eth_gecko.c]: It seems that currently all the existing drivers just have `eth_stats_update_errors_rx(context->iface);` to update the RX error count. So here you could just uncomment this line and it should work just fine. Just add `#include <ethernet/eth_stats.h>` at the beginning of the file. Currently all the other Ethernet stats are updated by the generic Ethernet L2 driver.

Comment (Oanerer) [reply drivers/ethernet/eth_gecko.c]: I see, thank you! I've uncommented `eth_stats_update_errors_rx(context->iface);` and included `<ethernet/eth_stats.h>`.

**Review by jukkar — APPROVED** [review]
+1 to Ethernet driver

**Comment by jhedberg** [comment]
@Oanerer you'll need to address the failed Codeowners and Shippable checks before this can be merged. Any update on this?

**Comment by Oanerer** [comment]
> 
> 
> @Oanerer you'll need to address the failed Codeowners and Shippable checks before this can be merged. Any update on this?
Thanks for asking!

1. The checkpatch warnings are mostly lines over 80 characters, but I don't know if that is an issue?
2. Same question for the `WARNING:UNDOCUMENTED_DT_STRING` warning; is that an issue? If so, how should I proceed?
3. Shippable fails probably because this PR depends on PR [#2](https://github.com/zephyrproject-rtos/hal_silabs/pull/2) in `hal_silabs`. If someone can merge that, I will also add a `west.yml` entry to that to this PR.
4. I will add a `CODEOWNERS` entry for the files along with the `west.yml` entry.

**Comment by jhedberg** [comment]
@Oanerer 1 and 2 are probably not blockers. Thanks for letting me know about the dependent hal_silabs PR (I should have noticed the reference to it in the description).

**Comment by jhedberg** [comment]
@Oanerer I've gone and merged the HAL PR now

**Comment by Oanerer** [comment]
> 
> 
> @Oanerer I've gone and merged the HAL PR now

Much appreciated! I updated `west.yml` accordingly and added an entry to `CODEOWNERS`. Hopefully shippable passes now also.

**Review by jhedberg — CHANGES_REQUESTED** [review]
Some minor cosmetic issues I noticed, which you may want to fix

File: drivers/ethernet/eth_gecko_priv.h
Comment (Oanerer) [reply drivers/ethernet/eth_gecko_priv.h]: Sure, removed them. 

As a force of habit I tend to add them as I personally rather have some redundant ones than one too little that causes a weird bug.

File: drivers/ethernet/eth_gecko_priv.h
Comment (Oanerer) [reply drivers/ethernet/eth_gecko_priv.h]: Modified these to the BIT() version and thanks for pointing it it's existence!

File: drivers/ethernet/eth_gecko_priv.h
Comment (Oanerer) [reply drivers/ethernet/eth_gecko_priv.h]: Done for 1 bit only values, left `ETH_RX_VLAN_PRIORITY` and `ETH_RX_OFFSET` intact.

File: drivers/ethernet/eth_gecko_priv.h
Comment (Oanerer) [reply drivers/ethernet/eth_gecko_priv.h]: I went for the latter approach, using the `eth` base address as the parameter for the macro.

File: drivers/ethernet/phy_gecko.c
Comment (Oanerer) [reply drivers/ethernet/phy_gecko.c]: Removed

**Comment by Oanerer** [comment]
> 
> 
> Some minor cosmetic issues I noticed, which you may want to fix

Thank you, the suggested changes have been implemented.

**Review by jhedberg — APPROVED** [review]
Looks ok to me, but it'd be good to get some more reviews from the assigned reviewers.

File: boards/arm/efm32gg_stk3701a/Kconfig
Comment (ioannisg) [inline boards/arm/efm32gg_stk3701a/Kconfig]: nit; can you remove this comment line?

Comment (Oanerer) [reply boards/arm/efm32gg_stk3701a/Kconfig]: Minor nit fixed, thanks

**Comment by Oanerer** [comment]
So... out of the blue Shippable fails on `SOC_FAMILY_EXX32` which still seems present in master and used by all other Gecko boards! How to proceed??

`error: SOC_FAMILY_EXX32 (defined at soc/arm/silabs_exx32/Kconfig:5) is assigned in a configuration file,
but is not directly user-configurable (has no prompt). It gets its value indirectly from other
symbols. See http://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_SOC_FAMILY_EXX32.html
and/or look up SOC_FAMILY_EXX32 in the menuconfig/guiconfig interface. The Application Development
Primer, Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual
might be helpful too.
`

**Review by ioannisg — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
